lib/cucumber/filters/retry.rb

```diff
@@ -0,0 +1,32 @@
+require 'cucumber/core/filter'
+require 'cucumber/running_test_case'
+require 'cucumber/events/bus'
+require 'cucumber/events/after_test_case'
+
+module Cucumber
+  module Filters
+    class Retry < Core::Filter.new(:configuration)
+
+      def test_case(test_case)
+        configuration.on_event(:after_test_case) do |event|
+          next unless retry_required?(test_case, event)
+
+          test_case_counts[test_case] += 1
+          event.test_case.describe_to(receiver)
+        end
+
+        super
+      end
+
+      private
+
+      def retry_required?(test_case, event)
+        event.test_case == test_case && event.result.failed? && test_case_counts[test_case] < configuration.retry_attempts
```

Comment:
This does not work, since the test_case you pass on is never the same that are in the `after_test_case` event. Filters like [activate_steps](https://github.com/cucumber/cucumber-ruby/blob/master/lib/cucumber/filters/activate_steps.rb#L22) call [Case#with_steps](https://github.com/cucumber/cucumber-ruby-core/blob/master/lib/cucumber/core/test/case.rb#L41) which creates a new test case object.